nodebuilder/header/data_commitment.go

```diff
@@ -0,0 +1,242 @@
+type DataRootTupleInclusionProof *merkle.Proof
+
+// padBytes Pad bytes to given length
+func padBytes(byt []byte, length int) ([]byte, error) {
+	l := len(byt)
+	if l > length {
+		return nil, fmt.Errorf(
+			"cannot pad bytes because length of bytes array: %d is greater than given length: %d",
+			l,
+			length,
+		)
+	}
+	if l == length {
+		return byt, nil
+	}
+	tmp := make([]byte, length)
+	copy(tmp[length-l:], byt)
+	return tmp, nil
+}
+
+// To32PaddedHexBytes takes a number and returns its hex representation padded to 32 bytes.
+// Used to mimic the result of `abi.encode(number)` in Ethereum.
+func To32PaddedHexBytes(number uint64) ([]byte, error) {
+	hexRepresentation := strconv.FormatUint(number, 16)
+	// Make sure hex representation has even length.
+	// The `strconv.FormatUint` can return odd length hex encodings.
+	// For example, `strconv.FormatUint(10, 16)` returns `a`.
+	// Thus, we need to pad it.
+	if len(hexRepresentation)%2 == 1 {
+		hexRepresentation = "0" + hexRepresentation
+	}
+	hexBytes, hexErr := hex.DecodeString(hexRepresentation)
+	if hexErr != nil {
+		return nil, hexErr
+	}
+	paddedBytes, padErr := padBytes(hexBytes, 32)
+	if padErr != nil {
+		return nil, padErr
+	}
+	return paddedBytes, nil
+}
+
+// DataRootTuple contains the data that will be used to create the QGB commitments.
+// The commitments will be signed by orchestrators and submitted to an EVM chain via a relayer.
+// For more information:
+// https://github.com/celestiaorg/quantum-gravity-bridge/blob/master/src/DataRootTuple.sol
+type DataRootTuple struct {
+	height   uint64
+	dataRoot [32]byte
+}
+
+// EncodeDataRootTuple takes a height and a data root, and returns the equivalent of
+// `abi.encode(...)` in Ethereum.
+// The encoded type is a DataRootTuple, which has the following ABI:
+//
+//	{
+//	  "components":[
+//	     {
+//	        "internalType":"uint256",
+//	        "name":"height",
+//	        "type":"uint256"
+//	     },
+//	     {
+//	        "internalType":"bytes32",
+//	        "name":"dataRoot",
+//	        "type":"bytes32"
+//	     },
+//	     {
+//	        "internalType":"structDataRootTuple",
+//	        "name":"_tuple",
+//	        "type":"tuple"
+//	     }
+//	  ]
+//	}
+//
+// padding the hex representation of the height padded to 32 bytes concatenated to the data root.
+// For more information, refer to:
+// https://github.com/celestiaorg/blobstream-contracts/blob/master/src/DataRootTuple.sol
+func EncodeDataRootTuple(height uint64, dataRoot [32]byte) ([]byte, error) {
+	paddedHeight, err := To32PaddedHexBytes(height)
+	if err != nil {
+		return nil, err
+	}
+	return append(paddedHeight, dataRoot[:]...), nil
+}
+
+// dataCommitmentBlocksLimit The maximum number of blocks to be used to create a data commitment.
+// It's a local parameter to protect the API from creating unnecessarily large commitments.
+const dataCommitmentBlocksLimit = 10_000 // ~33 hours of blocks assuming 12-second blocks.
```

Comment:
We just need to set some limit on the API for nodes not to be spammed with creating large proofs unnecessarily.
Also, we could have it as a parameter, but I don't think it's worth doing it now since no one will ask for proofs that are larger than 2000 blocks

Reply:
Yes i understand the reasoning for it, I just think you should use the parameter we defined to calculate the approximate num blocks (`p2p.BlockTime`)

Reply:
I see, the 10_000 is just a certain number of blocks that we think is enough. You mean using `p2p.BlockTime` to calculate the corresponding period in the comment?

Reply:
[b8d7a44](https://github.com/celestiaorg/celestia-node/pull/3470/commits/b8d7a444198905245439dc8743ac26e6f6f9a72f)